clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3416,129 +3358,179 @@ void SYCLIntegrationHeader::emitFwdDecl(raw_ostream &O, const Decl *D,
 //   template <typename T> class MyTmplClass;
 //   template <typename T1, unsigned int N, typename ...T2> class SimpleVadd;
 //
-void SYCLIntegrationHeader::emitForwardClassDecls(
-    raw_ostream &O, QualType T, SourceLocation KernelLocation,
-    llvm::SmallPtrSetImpl<const void *> &Printed) {
+class SYCLFwdDeclEmitter
+    : public TypeVisitor<SYCLFwdDeclEmitter>,
+      public ConstTemplateArgumentVisitor<SYCLFwdDeclEmitter> {
+  using InnerTypeVisitor = TypeVisitor<SYCLFwdDeclEmitter>;
+  using InnerTemplArgVisitor = ConstTemplateArgumentVisitor<SYCLFwdDeclEmitter>;
+  raw_ostream &OS;
+  llvm::SmallPtrSetImpl<const void *> &Printed;
```

Comment:
not a fan of void ptr here... Can this be at least 'Decl'?

Reply:
Also, it seems that 'Printed' has the same lifetime as the SYCLFwdDeclEmitter, so we should just make this an internal implementation detail instead of part of the ctor and caller, unless there is a good reason for it.